btf_interp/src/lib.rs

```diff
@@ -1,51 +1,293 @@
-    pub fn new(size: Option<NonZeroUsize>, adjust_tape: Option<bool>) -> Self {
+    pub fn new(
+        program: &'a BrainFuckProgram,
+        size: Option<NonZeroUsize>,
+        adjust_tape: Option<bool>,
+    ) -> Self {
         VirtualMachine {
             tape: Vec::new(),
             tape_size: size.map(NonZeroUsize::get).unwrap_or(3000),
             adjust_tape: adjust_tape.unwrap_or(false),
-            pointer: 0,
+            head: 0,
+            program,
         }
     }
 
-    /// Add element into tape. If tape size exceeded, element is not going to be added, error message is shown.
-    pub fn push(&mut self, item: T) -> Result<(), String> {
-        if self.tape.len() >= self.tape_size {
-            if !self.adjust_tape {
-                return Err(format!("Max tape size of {} reached.", self.tape_size));
-            }
-            self.tape_size += 1;
-        }
-        self.tape.push(item);
-        Ok(())
-    }
-
     /// Interpreter BF program into human-readable format.
-    pub fn interpreter(&self, bf_program: &BrainFuckProgram) {
-        for instruction_position in bf_program.instructions() {
+    pub fn interpreter(&self) {
+        for instruction_position in self.program.instructions() {
             println!(
                 "[{}:{}",
-                bf_program.filename().display(),
+                self.program.filename().display(),
                 instruction_position
             );
         }
     }
+
+    /// Go to the next element in tape. If tape size exceeded, error message is shown.
+    pub fn next_element(&mut self) -> Result<(), VMError> {
+        if self.head + 1 == self.tape_size {
+            let instruction = &self.program.instructions()[self.head];
+            return Err(VMError::NextElementNotReachable {
+                line: instruction.line(),
+                position: instruction.position(),
+            });
+        }
+        self.head += 1;
+        Ok(())
+    }
+
+    /// Go to the previous element in tape. If it is the first element, error message is shown.
+    pub fn previous_element(&mut self) -> Result<(), VMError> {
+        if self.head == 0 {
+            let instruction = &self.program.instructions()[self.head];
+            return Err(VMError::PreviousElementNotReachanble {
+                line: instruction.line(),
+                position: instruction.position(),
+            });
+        }
+        self.head -= 1;
+        Ok(())
+    }
+
+    /// Add 1 to the element where head is pointing to.
+    pub fn wrapped_add(&mut self) -> Result<(), u8> {
+        self.tape[self.head].wrapping_increment();
+        Ok(())
+    }
+
+    /// Substract 1 to the element where head is pointing to.
+    pub fn wrapped_sub(&mut self) -> Result<(), u8> {
+        self.tape[self.head].wrapping_decrement();
+        Ok(())
+    }
+
+    /// Basic IO read.
+    pub fn read(&mut self, reader: &mut impl Read) -> Result<(), VMError> {
```

Comment:
I have added basic IO read and write functions, but I am not sure that I understand how they can be used in the current implementation. Or it is a preparation for the next (final) iteration?